### DISCUSSION TITLE
Rust: Restart with a countdown

### DISCUSSION THREAD

**Discussion by xibalbus** [GameServerManagers/LinuxGSM#3429] [Q&A] [answered]
I'd like to use cron to schedule a restart at 5am, but I'd also like to give users a period of notice in Global chat.

Such as "Server restarting in 30 minutes".
"Server restarting in 15 minutes"
"Server restarting in 5 minutes."
* server restarts *

What is the best way to achieve this with LGSM?

**Comment by xox-Mia-xoxo** [comment]
here are already suggestions for ideas here?

**Comment by visdmin** [comment, marked as answer]
This issue is puzzling me, I am trying to find solution but cant promise anything.
I am researching following setup that runs from single bash script (can be used with cronjob):
- Uses command line rcon / web-rcon client to send command to rust server (`global.say "update in x minutes"` or some plugin command)
- Waits for the given period
- Stops the rust server
- Checks for updates & mods updates
- Start the rust server

I have found few command line rcon clients, but looking for better ones.
If i am able to find easy to install rcon command line client, i will create bash script that uses that to achieve functionality above.

-----
### Temporary solution ?

But here is something that you could maybe use until there is actual solution for your problem:

### Rust server plugin to create timed chat messages:
There is plugin that can be found from uMod that allows you to schedule execution of rust server commands:
- https://umod.org/plugins/timed-execute

Idea:
- Use `timed-execute` plugin to print the info messages for the players
- You server `cronjob` to `restart` OR even `update` the server.

**PLEASE NOTE:**
- Update script has been tested, but i have not tested the rust plugin that I used in this example.
- There could be problems related to timing of the cronjob and rust plugin commands.
- Complete example setup is untested.

Using the `timed-execute` plugin with the following config to print the info messages:
This example config would send chat messages for updates that happen every day `12:31` and `00:31`.
```json
{
  "EnableInGameTime-Timer": false,
  "EnableRealTime-Timer": true,
  "EnableTimerOnce": false,
  "EnableTimerRepeat": false,
  "InGameTime-Timer": {},
  "RealTime-Timer": {
    "12:00:00": "global.say Server restarting in 30 minutes !",
    "12:15:00": "global.say Server restarting in 15 minutes !",
    "12:20:00": "global.say Server restarting in 10 minutes !",
    "12:25:00": "global.say Server restarting in 5 minutes !",
    "12:29:30": "global.say Server restarting in 30 seconds !",
    "00:00:00": "global.say Server restarting in 30 minutes !",
    "00:15:00": "global.say Server restarting in 15 minutes !",
    "00:20:00": "global.say Server restarting in 10 minutes !",
    "00:25:00": "global.say Server restarting in 5 minutes !",
    "00:29:30": "global.say Server restarting in 30 seconds !",
  },
  "TimerOnce": {},
  "TimerRepeat": {}
}
```

### Using cronjob to restart rust server:
1. Setup crontab to executes the restart command `12:31` and `00:31` every day.
    ```shell
    crontab -e
    ```
    - Add following lines:
    ```
    31 12 * * * /opt/your-path/to/rustserver restart
    31 00 * * * /opt/your-path/to/rustserver restart
    ```
    - Save and exit.

This will setup automation that executes the `./rustserver restart` command by the given periods.
Calculator to create your own crontab timings: https://crontab.guru/#31_00_*_*_*
LGSM crontab documentation: https://docs.linuxgsm.com/configuration/cronjobs#crontab 

2. If you want to remove this automation:
    ```shell
    crontab -e
    ```
    - Remove the following lines:
    ```
    31 12 * * * /opt/your-path/to/rustserver restart
    31 00 * * * /opt/your-path/to/rustserver restart
    ```
    - Save and exit.

----

### Using cronjob to run script that updates the server:
**PLEASE NOTE:** All the following actions are done with the same server user you are using to run the rustserver.
1. Upload the script to your lgsm installation directory and make it executable:
    - I use `nano` text editor to copy-and-paste the script to my server, please use what ever method that suits you:
    ```shell
    nano /home/lgsm-daemon/rust_update_restart.sh
    ```
    - I copy paste the script into the newly created file
    - I change the `RUST_SERVER_PATH` variable to point to my `rustserver`:
    - I change the `UPDATE_MODS` variable to `true` or `false` if i want to run `mods-update`
    ```log
    RUST_SERVER_PATH="/opt/my-custom-directory/rust-server/rustserver"
    ```
    - save and exit with `CTRL + X` -> `Y` -> `ENTER`

2.  Run the script manually to test the setup
    ```shell
    ./rust_update_restart.sh
    ```
    - or 
    ```shell
    /opt/your-path/to/rust_update_restart.sh
    ```
3. Setup crontab to executes the update script, following rust the update `12:31` and `00:31` every day.
    ```shell
    crontab -e
    ```
    - Add following lines:
    ```
    31 12 * * * /opt/your-path/to/rust_update_restart.sh
    31 00 * * * /opt/your-path/to/rust_update_restart.sh
    ```
    - Save and exit.

Calculator to create your own crontab timings: https://crontab.guru/#31_00_*_*_*
LGSM crontab documentation: https://docs.linuxgsm.com/configuration/cronjobs#crontab 

4. If you want to remove the updating automation:
3. Setup crontab to executes the update script, following rust the update `12:31` and `00:31` every day.
    ```shell
    crontab -e
    ```
    - Remove the following lines:
    ```
    31 12 * * * /opt/your-path/to/rust_update_restart.sh
    31 00 * * * /opt/your-path/to/rust_update_restart.sh
    ```
    - Save and exit.

### Theory:
The setup described above should be able to give you:
- Automatic timed warning for players about the upcoming server update.
- Automatic server update.

Please let me know if anyone tries the described setup, or if you need any help !

----

`rust_update_restart.sh`: https://gist.github.com/visdmin/22916bae2500d052b0a55296b34dc1a3
